shared/chat/conversation/list-area/normal/index.native.js

```diff
@@ -1,57 +1,66 @@
 // @flow
 import * as React from 'react'
-import * as Types from '../../../../constants/types/chat2'
 import Message from '../../messages'
+import SpecialTopMessage from '../../messages/special-top-message'
+import SpecialBottomMessage from '../../messages/special-bottom-message'
 import {Box, NativeVirtualizedList, ErrorBoundary} from '../../../../common-adapters/index.native'
 
 import type {Props} from '.'
 
-const blankOrdinal = Types.numberToOrdinal(0)
-type State = {
-  calledLoadMoreDueToOrdinal: ?Types.Ordinal,
-}
+class ConversationList extends React.PureComponent<Props> {
+  _renderItem = ({index, item}) => {
+    if (item === 'specialTop') {
+      return <SpecialTopMessage conversationIDKey={this.props.conversationIDKey} measure={null} />
+    } else if (item === 'specialBottom') {
+      return <SpecialBottomMessage conversationIDKey={this.props.conversationIDKey} measure={null} />
+    } else {
+      const ordinalIndex = item
+      const ordinal = this.props.messageOrdinals.get(ordinalIndex)
+      const prevOrdinal = ordinalIndex > 0 ? this.props.messageOrdinals.get(ordinalIndex - 1) : null
 
-class ConversationList extends React.PureComponent<Props, State> {
-  state = {
-    calledLoadMoreDueToOrdinal: null,
+      return (
+        <Message
+          ordinal={ordinal}
+          previous={prevOrdinal}
+          measure={null}
+          conversationIDKey={this.props.conversationIDKey}
+        />
+      )
+    }
   }
 
-  _renderItem = ({index}) => {
-    const i = this.props.hasExtraRow ? index - 1 : index
-    const ordinal = i < 0 ? blankOrdinal : this.props.messageOrdinals.get(i, blankOrdinal)
-    const prevOrdinal = this.props.messageOrdinals.get(i + 1, blankOrdinal)
-    return (
-      <Message
-        ordinal={ordinal}
-        previous={prevOrdinal}
-        conversationIDKey={this.props.conversationIDKey}
-        measure={null}
-      />
-    )
+  _getItem = (messageOrdinals, index) => {
+    // Note we invert our list so we need to feed it things in the reverse order. We just invert the index
+    // vs reversing the items to speed things up
+    const itemCountIncludingSpecial = this._getItemCount(messageOrdinals)
+    if (index === itemCountIncludingSpecial - 1) {
+      return 'specialTop'
+    } else if (index === 0) {
+      return 'specialBottom'
+    }
+
+    // return ordinalIndex
```

Comment:
maybe this comment can mention that this return is doing the reverse calculation mentioned at L33